orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -1101,6 +1101,38 @@ IMPLEMENT_GRADIENT_BUILDER(GetReduceLogSumExpGradient) {
   return result;
 }
 
+IMPLEMENT_GRADIENT_BUILDER(GetReduceL2Gradient) {
+  std::vector<NodeDef> result;
+  auto attributes = SrcNodeAttributes();
+  bool keepdims = true;
+  if (attributes.find("keepdims") != attributes.end() &&
+      attributes.at("keepdims").has_i()) {
+    keepdims = static_cast<bool>(attributes.at("keepdims").i());
+  }
+
+  ArgDef y_arg_def = O(0);
+  ArgDef dy_arg_def = GO(0);
+  if (!keepdims && attributes.find("axes") != attributes.end()) {
+    std::vector<int64_t> axes_values = RetrieveValues<int64_t>(attributes.at("axes"));
+    y_arg_def = IA("Unsqueezed_Y");
+    dy_arg_def = IA("Unsqueezed_dY");
+    result.emplace_back(NodeDef("Unsqueeze", {O(0)}, {y_arg_def}, {MakeAttribute("axes", axes_values)}));
+    result.emplace_back(NodeDef("Unsqueeze", {GO(0)}, {dy_arg_def}, {MakeAttribute("axes", axes_values)}));
+  }
+
+  result.emplace_back(NodeDef("Div", {dy_arg_def, y_arg_def}, {IA("Scaled_dY")}));
```

Comment:
Swap order between "Div" and "Unsqueezes"? 
so you can potentially perform Div on a smaller tensor and save one Unsqueeze node.

Reply:
The element size is same before and after "Unsqueeze", but yes we can save one "Unsqueeze".